clinica/pipelines/statistics_surface/_model.py

```diff
@@ -0,0 +1,728 @@
+import abc
+import warnings
+from functools import reduce
+from os import PathLike
+from pathlib import Path
+from string import Template
+from typing import Callable, Dict, List, Optional, Union
+
+import numpy as np
+import pandas as pd
+from brainstat.stats.SLM import SLM
+from brainstat.stats.terms import FixedEffect
+from nilearn.surface import Mesh
+
+from clinica.utils.stream import cprint
+
+DEFAULT_THRESHOLD_UNCORRECTED_P_VALUE = 0.001
+DEFAULT_THRESHOLD_CORRECTED_P_VALUE = 0.05
+DEFAULT_CLUSTER_THRESHOLD = 0.001
+MISSING_TERM_ERROR_MSG = Template(
+    "Term ${term} from the design matrix is not in the columns of the "
+    "provided TSV file. Please make sure that there is no typo."
+)
+
+
+def _print_clusters(model: SLM, threshold: float) -> None:
+    """This function prints the results related to total number
+    of clusters, as well as the significative clusters.
+
+    Parameters
+    ----------
+    model : Fitted SLM model.
+    threshold : Cluster defining threshold.
+    """
+    cprint("#" * 40)
+    cprint("After correction (Cluster-wise Correction for Multiple Comparisons): ")
+    df = model.P["clus"][0]
+    cprint(df)
+    cprint(f"Clusters found: {len(df)}")
+    cprint(
+        f"Significative clusters (after correction): {len(df[df['P'] <= threshold])}"
+    )
+
+
+def _is_categorical(df: pd.DataFrame, column: str) -> bool:
```

Comment:
This function deals with two responsibilities, figure whether a column is categorical and whether a column exist in the input DataFrame which raises an exception. The latter use case cannot be deduce from the function name nor the docstring.

Instead maybe changing the logic to enumerate categorical columns from an input DataFrame?

To test whether a column [is categorical](https://pandas.pydata.org/pandas-docs/stable/user_guide/categorical.html#equality-semantics):

```python
column_dtype = dataframe.dtypes[column_name]
is_categorical = column_type == "category"
```

Reply:
I agree for the responsibility issue. I made some changes in [1a1e9bf](https://github.com/aramis-lab/clinica/pull/669/commits/1a1e9bf7d22797d5ceb7dbbd8445cecd85af3065) to improve this.

I might be missing something however for the way you suggest we test for categorical columns. I agree this would be the way to test it given a properly defined (meaning with dtypes set) DataFrame, but the DataFrame we are dealing with here comes from a brutal tsv read with no a priori information, typically on the types of the data we are reading. So the columns with categorical data will just have the `object` dtype. 
Does this make sense?

Reply:
> I agree for the responsibility issue. I made some changes in [1a1e9bf](https://github.com/aramis-lab/clinica/pull/669/commits/1a1e9bf7d22797d5ceb7dbbd8445cecd85af3065) to improve this.

That's better indeed 👍 

> DataFrame we are dealing with here comes from a brutal tsv read with no a priori information

So how do you figure whether a column is categorical or not from a raw string format without additional metadata, then? This is a very good question.

The implementation says anything whose dtype name does not start with "float". Which is a proxy for testing whether a column contains values which are non-float. So any column which dtype is either strings, objects or integers would be considered as categorical. Is it really the right logic, though?

Let's take the following example: if I have an age column successfully parsed as integers, then it's categorical. If it's parsed as floats, then it's not. Date-like columns would be considered categorical too. Also, `object` dtype is attributed when the `pandas` sniffers fail to detect a precise dtype from a column. Those columns may be categorical or malformed.

Should we decide to keep this logic, then it would be better to rename this filter to `is_nonfloat` rather than keep `is_categorical` in my opinion. Another option would be to enhance the reader to analyze columns which have been successfully converted as strings after `convert_dtype`, compute their histograms, apply a conservative heuristic between the range of different values over the total of lines and automatically convert them to a proper categorical dtype. It would be significantly more work though.

Reply:
The current implementation is legacy code I left as it was. I do agree it's far from perfect and might lead to spurious failures down the road as you explained above.

I liked the heuristic solution you proposed and went for an implementation in [66bc1c9](https://github.com/aramis-lab/clinica/pull/669/commits/66bc1c91b623a27601cefa542789842e98f185eb).
Let me know if this is what you had in mind !

One thing though is that using `convert_dtype` is currently breaking things in `Brainstat`.
I think this is a bug on their side, so I filled a bug report (https://github.com/MICA-MNI/BrainStat/issues/292), and opened a fix (https://github.com/MICA-MNI/BrainStat/pull/293).
We'll see how it goes... In the mean time, tests are broken, but pass with the proposed upstream fix.